dist/ExtUtils-ParseXS/t/002-more.t failing on Win32 (mingw64)

GH Actions tests on "Windows mingw64" are failing on multiple pull requests on different branches.  See, *e.g.,*

https://github.com/Perl/perl5/actions/runs/5447805747/jobs/9910433855?pr=21200

https://github.com/Perl/perl5/actions/runs/5448721934/jobs/9912356883?pr=21201

https://github.com/Perl/perl5/actions/runs/5447537853/jobs/9909828212?pr=21199

Sample failures:
```
...
../dist/ExtUtils-ParseXS/t/001-basic.t ............................... ok

#   Failed test 'No error message recorded, as expected'
#   at t/002-more.t line 65.
#          got: 'Can't load 'D:\a\perl5\perl5\dist\ExtUtils-ParseXS\t/XSMore.dll' for module XSMore: load_file:Invalid access to memory location at D:\a\perl5\perl5\lib/DynaLoader.pm line 206.
../dist/ExtUtils-ParseXS/t/002-more.t ................................ 
# '
Failed 26/33 subtests 
#     expected: ''

#   Failed test 'the BOOT keyword'
#   at t/002-more.t line 68.
#          got: undef
#     expected: '100'
Undefined subroutine &XSMore::include_ok called at t/002-more.t line 70.
# Looks like your test exited with 2 just after 9.
../dist/ExtUtils-ParseXS/t/003-usage.t ............................... ok
...

Test Summary Report
-------------------
../dist/ExtUtils-ParseXS/t/002-more.t                              (Wstat: 512 (exited 2) Tests: 9 Failed: 2)
  Failed tests:  7, 9
  Non-zero exit status: 2
  Parse errors: Bad plan.  You planned 33 tests but ran 9.
../dist/Net-Ping/t/450_service.t                                   (Wstat: 0 Tests: 26 Failed: 0)
  TODO passed:   9
../ext/IPC-Open3/t/IPC-Open3.t                                     (Wstat: 0 Tests: 45 Failed: 0)
  TODO passed:   25
Files=2767, Tests=1128665, 1839 wallclock secs (103.00 usr +  8.31 sys = 111.31 CPU)
Result: FAIL
Finished test run at Mon Jul  3 19:19:37 2023.
gmake: *** [GNUMakefile:1840: test] Error 2
Error: Process completed with exit code 2
```
I have no access to Windows, so I can't diagnose further.

--- a/.github/workflows/testsuite.yml
+++ b/.github/workflows/testsuite.yml
@@ -521,27 +521,18 @@ jobs:
     steps:
       - run: git config --global core.autocrlf false
       - uses: actions/checkout@v3
-      - name: Set up Perl build environment
-        run: |
-          # skip installing perl if it is already installed.
-          if (!(Test-Path "C:\strawberry\perl\bin")) {
-            choco install strawberryperl
-          }
-          echo @"
-          C:\strawberry\c\bin
-          C:\strawberry\perl\site\bin
-          C:\strawberry\perl\bin
-          "@ |
-            Out-File -FilePath $env:GITHUB_PATH -Encoding utf8 -Append
+      - name: Install mingw-64
+        uses: egor-tensin/setup-mingw@v2
       - name: Host perl -V
-        run: perl -V
+        run: |
+          perl -V
       - name: gcc --version
         run: gcc --version
       - name: Build
         shell: cmd
         run: |
           cd win32
-          gmake CCHOME=C:\strawberry\c CFG=Debug -f GNUMakefile -j2
+          gmake CCHOME=C:\ProgramData\chocolatey\lib\mingw\tools\install\mingw64 CFG=Debug -f GNUMakefile -j2
       - name: Show Config
         shell: cmd
         run: |
@@ -552,7 +543,7 @@ jobs:
         run: |
           cd win32
           set HARNESS_OPTIONS=j2
-          gmake CCHOME=C:\strawberry\c CFG=Debug -f GNUMakefile test
+          gmake CCHOME=C:\ProgramData\chocolatey\lib\mingw\tools\install\mingw64 CFG=Debug -f GNUMakefile test
 
   #                            _
   #   ___ _   _  __ ___      _(_)_ __

--- a/win32/config_sh.PL
+++ b/win32/config_sh.PL
@@ -323,10 +323,6 @@ if ($opt{cc} =~ /\bcl/ and $opt{ccversion} =~ /^(\d+)/) {
 	$opt{i_stdint} = 'define';
     }
     if ($ccversion >= 19) { # VC14+
-	$opt{stdio_base} = 'PERLIO_FILE_base(fp)';
-	$opt{stdio_bufsiz} = '(PERLIO_FILE_cnt(fp) + PERLIO_FILE_ptr(fp) - PERLIO_FILE_base(fp))';
-	$opt{stdio_cnt} = 'PERLIO_FILE_cnt(fp)';
-	$opt{stdio_ptr} = 'PERLIO_FILE_ptr(fp)';
 	$opt{i_stdbool} = 'define' unless $prebuilt;
     }
 }
@@ -342,14 +338,17 @@ elsif ($opt{cc} =~ /\bicl/) {
 	$opt{i_stdint} = 'define';
     }
     if ($num_ver =~ /^(\d+)/ && $1 >= 19) { # VC14+
-	$opt{stdio_base} = 'PERLIO_FILE_base(fp)';
-	$opt{stdio_bufsiz} = '(PERLIO_FILE_cnt(fp) + PERLIO_FILE_ptr(fp) - PERLIO_FILE_base(fp))';
-	$opt{stdio_cnt} = 'PERLIO_FILE_cnt(fp)';
-	$opt{stdio_ptr} = 'PERLIO_FILE_ptr(fp)';
 	$opt{i_stdbool} = 'define';
     }
     $opt{ar} ='xilib';
 }
+
+# win32.h always defines these, and those definitions
+# depend on the current CRT
+$opt{stdio_base} = 'PERLIO_FILE_base(fp)';
+$opt{stdio_bufsiz} = '(PERLIO_FILE_cnt(fp) + PERLIO_FILE_ptr(fp) - PERLIO_FILE_base(fp))';
+$opt{stdio_cnt} = 'PERLIO_FILE_cnt(fp)';
+$opt{stdio_ptr} = 'PERLIO_FILE_ptr(fp)';
 
 if ($opt{useithreads} eq 'define' && $opt{ccflags} =~ /-DPERL_IMPLICIT_SYS\b/) {
     $opt{d_pseudofork} = 'define';

--- a/win32/win32.h
+++ b/win32/win32.h
@@ -259,8 +259,10 @@ union PerlNan { unsigned __int64 __q; double __d; };
 extern const __declspec(selectany) union PerlNan __PL_nan_u = { 0x7FF8000000000000UI64 };
 #define NV_NAN ((NV)__PL_nan_u.__d)
 
+#endif /* ifdef _MSC_VER */
+
 /* The CRT was rewritten in VS2015. */
-#if _MSC_VER >= 1900
+#ifdef _UCRT
 
 /* No longer declared in stdio.h */
 EXTERN_C char *gets(char* buffer);
@@ -295,11 +297,7 @@ typedef struct
 #define PERLIO_FILE_flag(f) ((int)(((__crt_stdio_stream_data*)(f))->_flags))
 #define PERLIO_FILE_file(f) (*(int*)(&((__crt_stdio_stream_data*)(f))->_file))
 
-#endif
-
-#endif /* _MSC_VER */
-
-#if (!defined(_MSC_VER)) || (defined(_MSC_VER) && _MSC_VER < 1900)
+#else /* ifdef _UCRT */
 
 /* Note: PERLIO_FILE_ptr/base/cnt are not actually used for GCC or <VS2015
  * since FILE_ptr/base/cnt do the same thing anyway but it doesn't hurt to

